src/Workspaces/CSharp/Portable/LanguageServices/CSharpTypeInferenceService.TypeInferrer.cs

```diff
@@ -1821,108 +1809,71 @@ private static ITypeSymbol GetMemberType(ISymbol memberSymbol)
             private IEnumerable<TypeInferenceInfo> InferTypeInRefExpression(RefExpressionSyntax refExpression)
                 => InferTypes(refExpression);
 
-            private IEnumerable<TypeInferenceInfo> InferTypeForReturnStatement(ReturnStatementSyntax returnStatement, SyntaxToken? previousToken = null)
+            private ITypeSymbol UnwrapTaskLike(ITypeSymbol type, bool isAsync)
             {
-                InferTypeForReturnStatement(returnStatement, previousToken,
-                    out var isAsync, out var types);
-
-                if (!isAsync)
+                if (isAsync)
                 {
-                    return types;
-                }
+                    if (type.OriginalDefinition.Equals(this.Compilation.TaskOfTType()))
```

Comment:
:bulb: For task-like types (including `ValueTask<T>`), the type does not need to match `Task<T>`. For our purposes, we can treat it as the first generic argument of the type, or `void` if the type is not generic. It appears the original code did not handle these cases either, but considering the effort you put into eliminating the known bugs in this code it seems like something you may want to cover at the same time. 😄

Reply:
I am aware of this but I would prefer if this was fixed separately since it affects many other features as well, such as "use block body" and "introduce variable". In my opinion there should be one helper function to unwrap a task-like that they could all use.

Reply:
➡️  Moved to #30559

Reply:
Thanks